Sources/GRPC/GRPCClientChannelHandler.swift

```diff
@@ -0,0 +1,374 @@
+      // message for unary and client-streaming RPCs.
+      for message in messages {
+        context.fireChannelRead(self.wrapInboundOut(.response(message)))
+      }
+    case .failure(let error):
+      context.fireErrorCaught(error)
+    }
+  }
+}
+
+// MARK: - GRPCClientChannelHandler: Outbound
+extension GRPCClientChannelHandler: ChannelOutboundHandler {
+  public typealias OutboundIn = GRPCClientRequestPart2<Request>
+  public typealias OutboundOut = HTTP2Frame
+
+  public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+    switch self.unwrapOutboundIn(data) {
+    case .head(let requestHead):
+      // Feed the request into the state machine:
+      let result = self.stateMachine.sendRequestHeaders(
+        scheme: requestHead.scheme,
+        host: requestHead.host,
+        path: requestHead.path,
+        options: requestHead.options,
+        requestID: requestHead.requestID
+      )
+
+      switch result {
+      case .success(let headers):
+        // Great, we're clear to write the some headers. Create an appropriate frame and write it.
+        let frame = HTTP2Frame(streamID: self.streamID, payload: .headers(.init(headers: headers)))
+        context.write(self.wrapOutboundOut(frame), promise: promise)
+
+      case .failure(let sendRequestHeadersError):
+        switch sendRequestHeadersError {
+        case .invalidState:
+          // This is bad: we need to trigger an error and close.
+          promise?.fail(sendRequestHeadersError)
+          context.fireErrorCaught(GRPCError.client(.invalidState("unable to initiate RPC")))
+        }
+      }
+
+    case .request(let request):
+      // Feed the request message into the state machine:
+      let result = self.stateMachine.sendRequest(request, allocator: context.channel.allocator)
+      switch result {
+      case .success(let buffer):
+        // We're clear to send a message; wrap it up in an HTTP/2 frame.
+        let frame = HTTP2Frame(
+          streamID: self.streamID,
+          payload: .data(.init(data: .byteBuffer(buffer)))
+        )
+        context.write(self.wrapOutboundOut(frame), promise: promise)
+
+      case .failure(let writeError):
+        switch writeError {
+        case .cardinalityViolation:
+          // This is fine: we can ignore the request.
+          promise?.fail(writeError)
+
+        case .serializationFailed:
+          // This is bad: we need to trigger an error and close.
+          promise?.fail(writeError)
+          context.fireErrorCaught(GRPCError.client(.requestProtoSerializationFailure))
+
+        case .invalidState:
+          promise?.fail(writeError)
+          context.fireErrorCaught(GRPCError.client(.invalidState("unable to write message")))
+        }
+      }
+
+    case .end:
+      // Okay: can we close the request stream?
+      switch self.stateMachine.sendEndOfRequestStream() {
+      case .success(()):
```

Comment:
Then I'd just use `case .success:` here, and possibly make sure that the corresponding enum case has no associated values, if possible.

Reply:
At the moment we have that guarantee (that there are no other associated values) because we have switched over all possible cases here: if we use `case .success` then it would be possible for the associated value to change without use noticing. I'm still happy to change this though.

Reply:
Sorry, I didn't look at the type of the result of `self.stateMachine.sendEndOfRequestStream()`. You are right, `.success(())` is better.